ddtrace/internal/injection.py

```diff
@@ -0,0 +1,163 @@
+
+
+HookType = Callable[[Any], Any]
+HookInfoType = Tuple[HookType, int, Any]
+
+HOOK_ARG_PREFIX = "_hook_arg"
+
+
+class InvalidLine(Exception):
+    """
+    Raised when trying to inject a hook on an invalid line, e.g. a comment or a blank line.
+    """
+
+
+def _inject_hook(code, hook, lineno, arg):
+    # type: (Bytecode, HookType, int, Any) -> None
+    """Inject a hook at the given line number inside an abstract code object.
+
+    The hook is called with the given argument, which is also used as an
+    identifier for the hook itself. This should be kept in case the hook needs
+    to be removed.
+    """
+    for i, instr in enumerate(code):
+        try:
+            if instr.lineno == lineno:
+                # gotcha!
+                break
+        except AttributeError:
+            # pseudo-instruction (e.g. label)
+            pass
+    else:
+        raise InvalidLine("Line %d does not exist or is either blank or a comment" % lineno)
+
+    # DEV: This is the bytecode equivalent of
+    # >>> hook(arg)
+    # Additionally, we must discard the return value (top of the stack) to
+    # restore the stack to the state prior to the call.
+    code[i:i] = Bytecode(
+        [
+            Instr("LOAD_CONST", hook, lineno=lineno),
+            Instr("LOAD_CONST", arg, lineno=lineno),
+            Instr("CALL_FUNCTION", 1, lineno=lineno),
+            Instr("POP_TOP", lineno=lineno),
+        ]
+    )
+
+
+def _eject_hook(code, hook, line, arg):
+    # type: (Bytecode, HookType, int, Any) -> None
+    """Eject a hook from the abstract code object at the given line number.
+
+    The hook is identified by its argument. This ensures that only the right
+    hook is ejected.
+    """
+    for i, instr in enumerate(code):
+        try:
+            # DEV: We look at the expected opcode pattern to match the injected
+            # hook and we also test for the expected opcode arguments
+            if (
+                instr.lineno == line
+                and code[i].arg is hook
+                and code[i + 1].arg is arg
+                and [code[_].name for _ in range(i, i + 4)] == ["LOAD_CONST", "LOAD_CONST", "CALL_FUNCTION", "POP_TOP"]
+            ):
+                # gotcha!
+                break
+        except AttributeError:
+            # pseudo-instruction (e.g. label)
+            pass
+        except IndexError:
+            pass
+    else:
+        raise InvalidLine("Line %d does not contain a hook" % line)
+
+    del code[i : i + 4]
+
+
+def _function_with_new_code(f, abstract_code):
+    f.__code__ = abstract_code.to_code()
+    return f
+
+
+def inject_hooks(f, hooks):
+    # type: (FunctionType, List[HookInfoType]) -> List[HookInfoType]
+    """Bulk-inject a list of hooks into a function.
+
+    Hooks are specified via a list of tuples, where each tuple contains the hook
+    itself, the line number and the identifying argument passed to the hook.
+
+    Returns the list of hooks that failed to be injected.
```

Comment:
🤔 should we still raise? but just raise at the end with a list of the failed hooks attached to the exception?

I don't think it is that big of a deal, just comparing against `inject_hook` which would raise an exception

(not asking for a change, feel free to resolve)

Reply:
I have thought about this and opted for the current approach to avoid yet another custom exception which would require its own documentation for how to retrieve the list etc... WDYT?

Reply:
yeah, I had the same thought. okie dokie, as-is is good!